frontend/src/helpers/helpers.ts

```diff
@@ -1,3 +1,18 @@
+import { dequal } from 'dequal';
+import { useRef, useMemo } from 'preact/hooks';
+
 export const doNothing = () => {
   /* nothing to do */
 };
+
+export function useDeepCompareMemoize<T>(value: T): T {
+  const ref = useRef<T>(value);
+  const signalRef = useRef<number>(0);
+
+  if (!dequal(value, ref.current)) {
+    ref.current = value;
+    signalRef.current += 1;
+  }
+
+  return useMemo(() => ref.current, [signalRef.current]);
```

Comment:
While this is definitely an intriguing technique, I also find it slightly hard to follow and a rather unusual use for `useMemo`. Especially since `ref` and `signalRef` only change together, so I would expect the `useMemo` to not make a difference?
Couldn't we just directly use the `ref`, i.e.

```js
export function useDeepCompareMemoize<T>(value: T): T {
   const ref = useRef<T>(value);

   if (!dequal(value, ref.current)) {
     ref.current = value;
   }

   return ref.current;
}
```

or am I overlooking something here?

Reply:
Maybe we are, but I don't see it either. I was inspired by the implementation of https://github.com/kentcdodds/use-deep-compare-effect/blob/main/src/index.ts#L32 without questioning the actual implementation. Going by the history of this implementation, the `signalRef` was first introduced as a speedup, but later on the return value also change. I would question if the speedup aspect is actually still present.

In order to keep it simple here, I have gone with the standard implementation that you suggested. Moreover, I added test coverage.